daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala

```diff
@@ -92,9 +95,157 @@ object TransactionConversions {
+                        // we currently relax this check by widening the latter set, treating a node the
+                        // same regardless of whether it was under a rollback node or not.
+                        // Computed outputs that are not actual outputs can be safely trimmed: examples
+                        // are transient contracts and, now, potentially also outputs from nodes under
+                        // rollback.
+                        Right(contractIdsOrKeys)
+
+                      case NodeTypeCase.CREATE =>
+                        val protoCreate = node.getCreate
+                        for {
+                          newContractIdsOrKeys <- TransactionCoder
+                            .nodeKey(nodeVersion, protoCreate)
+                            .map {
+                              case Some(key) => contractIdsOrKeys + ContractIdOrKey.Key(key)
+                              case None => contractIdsOrKeys
+                            }
+                          contractId <- ValueCoder.CidDecoder
+                            .decode(protoCreate.getContractIdStruct)
+                        } yield newContractIdsOrKeys + ContractIdOrKey.Id(contractId)
+
+                      case NodeTypeCase.EXERCISE =>
+                        val protoExercise = node.getExercise
+                        for {
+                          newContractIdsOrKeys <- TransactionCoder
+                            .nodeKey(nodeVersion, protoExercise)
+                            .map {
+                              case Some(key) => contractIdsOrKeys + ContractIdOrKey.Key(key)
+                              case None => contractIdsOrKeys
+                            }
+                          contractId <- ValueCoder.CidDecoder
+                            .decode(protoExercise.getContractIdStruct)
+                        } yield newContractIdsOrKeys + ContractIdOrKey.Id(contractId)
+
+                      case NodeTypeCase.FETCH =>
+                        // A fetch may cause a divulgence, which is why the target contract is a potential output.
+                        ValueCoder.CidDecoder.decode(node.getFetch.getContractIdStruct).map {
+                          contractId => contractIdsOrKeys + ContractIdOrKey.Id(contractId)
+                        }
+
+                      case NodeTypeCase.LOOKUP_BY_KEY =>
+                        // Contract state only modified on divulgence, in which case we'll have a fetch node,
+                        // so no outputs from lookup node.
+                        Right(contractIdsOrKeys)
+
+                      case NodeTypeCase.NODETYPE_NOT_SET =>
+                        Left(ValueCoder.DecodeError("NODETYPE_NOT_SET not supported"))
+                    }
+                  }
+                case (Left(error), _) => Left(error)
+              }
+          }
+          .left
+          .map(ConversionError.DecodeError)
+    }
+
+  def filterCreateAndExerciseNodes(
```

Comment:
The name here doesn’t tell you if we’re keeping the creates & exercise (and only those) or if those are the ones that are filtered out. I’d suggest renaming it (e.g. `keepCreateAndExerciseNodes`, `) and adding a doc comment.

Reply:
The idea was to follow the same idea as in Scala but you're right, maybe it's not clear. Changed.